### PR TITLE
Ensure System.Uri consistently uses the correct RFC 3986 reserved character set

### PR DESCRIPTION
**Pull request by rmkerr**
There is an issue with System.Uri where certain percent encoded characters that should be left as-is are decoded, but only when the URI also contains Unicode or an encoded non-reserved symbol. This issue was heavily reported in .NET Framework, also appears to repro in CoreFx. Increased usage of Unicode has caused a recent rise in the number of reports.  

Here are the conditions under which the unexpected behavior will occur: 

- The URI contains the percent encoded form of any of the following reserved characters:  :, ', (, ), ! or *.
- The URI contains a Unicode or encoded non-reserved character. 

The incorrect behavior is to decode the reserved characters, but only when both of the above are true. This is the wrong behavior not only because it is inconsistent (depending on the presence of Unicode or encoded non-reserved), but also because it goes against the spec:  
```
   Percent-encoding a reserved character, or decoding a percent-encoded octet 
   that corresponds to a reserved character, will change how the URI is 
   interpreted by most applications.  Thus, characters in the reserved 
   set are protected from normalization and are therefore safe to be 
   used by scheme-specific and producer-specific algorithms for 
   delimiting data subcomponents within a URI. 
```
Uri.ToString is the only exception to the repro conditions above. The old ToString behavior was to always decode all characters except ?, #, % and the RFC 2396 reserved set. The new behavior is updates the reserved set to match RFC 3986. The RFC 2396 reserved set is a strict subset of the RFC 3986 set, and so no new characters will be decoded. This change makes ToString safer to use, as it can no longer change the semantics of a URI that uses ':' or the RFC 3986 reserved set.

The source of the issue was an update to the URI code to match the following change in the URI spec, where the new version adds a handful of reserved characters: 
```
RFC 2396 
reserved    = ";" | "/" | "?" | ":" | "@" | "&" | "=" | "+" | 
              "$" | "," 
RFC 3986/3987 
reserved    = gen-delims / sub-delims 
gen-delims  = ":" / "/" / "?" / "#" / "[" / "]" / "@" 
sub-delims  = "!" / "$" / "&" / "'" / "(" / ")" 
            / "*" / "+" / "," / ";" / "=" 
```
When System.Uri was updated to follow RFC 3986, some codepaths missed the update to the new reserved set. Thus the added reserved symbols are not considered reserved in some circumstances. In addition, one function is also missing ':' from the original reserved set.

This has been fixed in .Net Framework 4.7.2. This fix updates places where the RFC 2396 set was still being used, and adds tests to ensure this behavior is consistent. It also retains the code used for quirking in .Net Framework, but always enables the new behavior instead of checking an AppContextSwitch.

### CODE REVIEW

File: src/System.Private.Uri/src/System/UriSyntax.cs [hunk 1]
```diff
@@ -100,6 +100,14 @@ public abstract partial class UriParser
 
         internal static UriParser VsMacrosUri;
 
+        internal static bool DontEnableStrictRFC3986ReservedCharacterSets
+        {
+            // In .NET Framework this would test agains an AppContextSwitch. Since this is a potentially
```
Comment (caesar1995) [inline src/System.Private.Uri/src/System/UriSyntax.cs]: Nit: test against

File: src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs [hunk 2]
```diff
@@ -6,5 +6,11 @@ namespace System
 {
     public class UriParser
     {
+    	internal static bool DontEnableStrictRFC3986ReservedCharacterSets
+        {
```
Comment (caesar1995) [inline src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs]: Nit: spacing is not correct.

File: src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs [hunk 3]
```diff
@@ -6,5 +6,11 @@ namespace System
...
+            get {
```
Comment (caesar1995) [inline src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs]: Nit: put the bracket in next line.

File: src/System.Private.Uri/src/System/UriSyntax.cs [hunk 4]
```diff
@@ -100,6 +100,14 @@ public abstract partial class UriParser
...
+            // breaking change, we'll leave in the system used to disable it.
+            get {
```
Comment (caesar1995) [inline src/System.Private.Uri/src/System/UriSyntax.cs]: Nit: put the bracket in next line.

File: src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs (hunk 2 above)
Comment (DavidGoll) [inline src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs]: nit: tab char

Comment (DavidGoll) [reply src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs]: Will be fixed when you fix the tab char.

---
In reply to: [158154555](https://github.com/dotnet/corefx/pull/26006#discussion_r158154555) [](ancestors = 158154555)

File: src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Reflection;
+using System.Text;
+
+using Xunit;
+
+namespace System.PrivateUri.Tests
+{
```
Comment (DavidGoll) [inline src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs]: Do any of these tests need to be skipped for NETFX (pending onboarding of 4.7.2)?

Comment (rmkerr) [reply src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs]: All of them do! Thanks for the reminder.

File: src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs (hunk 2 above)
Comment (rmkerr) [reply src/System.Private.Uri/tests/UnitTests/Fakes/FakeUriParser.cs]: Fixed in the latest commit.

File: src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs (hunk 5 above)
Comment (rmkerr) [reply src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs]: This should now be fixed in the latest commit.

File: src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+            string path = "/%C3%88/";
+            for (int i = 0; i < RFC3986CompliantDecoding.GetLength(0); i++)
+            {
+                string sourceStr = baseUri + RFC3986CompliantDecoding[i, 0] + "@" + host + path;
+                uri = new Uri(sourceStr);
+                Assert.Equal(RFC3986CompliantDecoding[i, 1], uri.UserInfo);
+            }
+    }
+    }
+}
```
Comment (DavidGoll) [inline src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs]: nit: add a newline

Comment (rmkerr) [reply src/System.Private.Uri/tests/FunctionalTests/IriEncodingDecodingTests.cs]: Should be fixed now, as well as the incorrect indentation there on line 152.

**Pull request merged**